frontend/pages/upload.vue

```diff
@@ -0,0 +1,407 @@
+  isProcessing.value = true
+  startedProcessing.value = true
+  setAllStates(FileStatus.Waiting)
+  const startPortfolioUploadResp = await pactaClient.startPortfolioUpload({
+    items: fileStates.value.map((fileState: FileState) => ({
+      file_name: fileState.file.name,
+      // TODO(grady) consider adding file size here as a validation step.
+      // TODO(grady) consider adding file type validation here.
+      // TODO(grady) consider adding duplicate file checking here.
+    })),
+    holdings_date: {
+      time: holdingsDate.value.toISOString(),
+    },
+  }).catch(e => {
+    console.log('error starting upload', e, e.body)
+    if (e.body?.error_id) {
+      errorCode.value = e.body.error_id
+    } else {
+      errorCode.value = 'Unknown Error'
+    }
+    if (e.body?.message) {
+      errorMessage.value = e.body.message
+    } else {
+      errorMessage.value = 'An unexpected error occurred - please file a bug to help us fix it.'
+    }
+    setAllStates(FileStatus.Selected)
+  })
+  if (!startPortfolioUploadResp) {
+    isProcessing.value = false
+    return
+  }
+
+  setAllStates(FileStatus.Uploading)
+  const uploads = fileStatesWithDetail.value.map(async (fileState: FileStateDetail): Promise<void> => {
+    const { file, index } = fileState
+    const respItem = startPortfolioUploadResp.items.find((item) => item.file_name === file.name)
+    if (!respItem) {
+      throw new Error(`Could not find start item for file ${file.name} - something is probably wrong in the API bizlogic`)
+    }
+    fileStates.value[index].incompleteUploadId = respItem.incomplete_upload_id
+
+    await $axios({
+      method: 'PUT',
+      url: respItem.upload_url,
+      data: file,
+      headers: {
+        'Content-Type': file.type,
+        'x-ms-blob-type': 'BlockBlob',
+      },
+    }).then(() => {
+      fileStates.value[index].status = FileStatus.Uploaded
+    }).catch((e) => {
+      console.log('error uploading file', e)
+      fileStates.value[index].status = FileStatus.Error
+      fileStates.value[index].errorMessage = serializeError(e)
+      errorCode.value = 'Upload Failed'
+      errorMessage.value = 'One or more files could not be uploaded - please delete/resolve them and try again.'
+    })
+  })
+  await Promise.all(uploads)
+
+  let hadError = false
+  for (const fileState of fileStates.value) {
+    if (fileState.status === FileStatus.Uploaded) {
+      continue
+    } else if (fileState.status === FileStatus.Error) {
+      hadError = true
+    } else {
+      throw new Error(`Unexpected file state ${fileState.status}`)
+    }
+  }
+  if (!hadError) {
+    await doParsing()
+  }
+  isProcessing.value = false
+}
+
+const doParsing = async () => {
+  fileStates.value.forEach((_, i) => {
+    fileStates.value[i].status = FileStatus.Validating
+  })
+  await pactaClient.completePortfolioUpload({
+    items: fileStates.value.map((fileState) => ({
+      incomplete_upload_id: presentOrFileBug(fileState.incompleteUploadId),
```

Comment:
How do the `presentOrX` modal helpers function in a loop? Just want to make sure we don't just start savagely flashing the user's screen if 10 uploads fail or something.

Reply:
Since they throw, I assume they break the cycle of execution, and are caught at the higher level, which doesn't resume the execution of the thing that threw, just clears the error. Standard throw semantics no?

Reply:
Good call, I forgot they throw
![2023-12-09_10-30-38](https://github.com/RMI-PACTA/app/assets/2402010/efa4299f-7cf6-4898-8393-3f992533ca39)